enterprise/server/remote_execution/dirtools/dirtools.go

```diff
@@ -769,41 +817,54 @@ func (ff *BatchFileFetcher) bytestreamReadFiles(ctx context.Context, instanceNam
 	if len(fps) == 0 {
 		return nil
 	}
-	fp := fps[0]
-	var mode os.FileMode = 0644
-	if fp.FileNode.IsExecutable {
-		mode = 0755
-	}
-	f, err := os.OpenFile(fp.FullPath, os.O_RDWR|os.O_CREATE, mode)
-	if err != nil {
-		return err
-	}
-	resourceName := digest.NewResourceName(fp.FileNode.Digest, instanceName, rspb.CacheType_CAS, ff.digestFunction)
-	if ff.supportsCompression() {
-		resourceName.SetCompressor(repb.Compressor_ZSTD)
-	}
-	if err := cachetools.GetBlob(ctx, bsClient, resourceName, f); err != nil {
-		return err
-	}
+	fpInterface, err, _ := ff.group.Do(d.GetHash(), func() (interface{}, error) {
+		fp0 := fps[0]
+		var mode os.FileMode = 0644
+		if fp0.FileNode.IsExecutable {
+			mode = 0755
+		}
+		f, err := os.OpenFile(fp0.FullPath, os.O_RDWR|os.O_CREATE, mode)
+		if err != nil {
+			return nil, err
+		}
+		resourceName := digest.NewResourceName(fp0.FileNode.Digest, instanceName, rspb.CacheType_CAS, ff.digestFunction)
+		if ff.supportsCompression() {
+			resourceName.SetCompressor(repb.Compressor_ZSTD)
+		}
+		if err := cachetools.GetBlob(ctx, bsClient, resourceName, f); err != nil {
```

Comment:
any idea what would happen to the other actions if the first action to reach the group here gets canceled?

Reply:
Sure, done.